Proposal deposits in SPO voting stake

aniketd opened the pull request:
# Description

Resolves #4061 

- [x] Add tests
- [x] Adjust all involved code commentary
- [x] Address review comments

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

aniketd committed:
Add proposal-deposit to SPO voting stake calc.

* Add individualPoolStakeCoin field to IndividualPoolStake as numerator.
* Add pdTotal field to PoolDistr as denominator.
* Change calculatePoolDistr' to include the numerator and denominator.
* Update computeDRepDistr to include updating PoolDistr numerator and
denominator with proposal-deposits.
* Update spoAcceptedRatio to calculate in terms of the numerator/
denominator pair instead of Rational.

aniketd committed:
Imptests: prop-deposits counted in SPO voting stake

aniketd committed:
Clean up code and comments

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/PoolDistr.hs:
```diff
@@ -48,16 +54,21 @@ import NoThunks.Class (NoThunks (..))
 -- delegated to a registered stake pool.
 data IndividualPoolStake c = IndividualPoolStake
   { individualPoolStake :: !Rational
+  , individualPoolStakeCoin :: !(CompactForm Coin) -- TODO: Document
```
Avoid using types in the names of record fields. They are not very descriptive and cause problems when the type changes. Also in this case it is `CompactCoin`, not just `Coin`
```suggestion
  , -- | Total stake delegated to this pool
    individualTotalPoolStake :: !(CompactForm Coin)
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/PoolDistr.hs:
```diff
@@ -67,46 +78,70 @@ instance Crypto c => DecCBOR (IndividualPoolStake c) where
+toIndividualPoolStakePair indivPoolStake@(IndividualPoolStake _ _ _) =
   let IndividualPoolStake {..} = indivPoolStake
    in [ "individualPoolStake" .= individualPoolStake
+      , "individualPoolStakeCoin" .= individualPoolStakeCoin
```
Just so we don't forget to rename the JSON field too
```suggestion
      , "individualTotalPoolStake" .= individualTotalPoolStake
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/PoolDistr.hs:
```diff
@@ -67,46 +78,70 @@ instance Crypto c => DecCBOR (IndividualPoolStake c) where
-newtype PoolDistr c = PoolDistr
+data PoolDistr c = PoolDistr
   { unPoolDistr :: Map (KeyHash 'StakePool c) (IndividualPoolStake c)
+  , pdTotal :: !(CompactForm Coin)
```
```suggestion
  , -- | This is the total amount of ADA that has been delegated to all registered stake pools
    pdTotalActiveStake :: !(CompactForm Coin)
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/PoolDistr.hs:
```diff
@@ -67,46 +78,70 @@ instance Crypto c => DecCBOR (IndividualPoolStake c) where
 -- verification key) to 'IndividualPoolStake'.
-newtype PoolDistr c = PoolDistr
+data PoolDistr c = PoolDistr
   { unPoolDistr :: Map (KeyHash 'StakePool c) (IndividualPoolStake c)
```
It is no longer a newtype, we need to make the field strict
```suggestion
  { unPoolDistr :: !(Map (KeyHash 'StakePool c) (IndividualPoolStake c))
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/PoolDistr.hs:
```diff
@@ -48,16 +54,21 @@ import NoThunks.Class (NoThunks (..))
 -- delegated to a registered stake pool.
 data IndividualPoolStake c = IndividualPoolStake
   { individualPoolStake :: !Rational
```
```suggestion
  { -- | Pool stake distribution. This is a ratio of `individualTotalPoolStake`/`pdTotalActiveStake`
    individualPoolStake :: !Rational
```

aniketd replied:
I have updated the comment to reflect that proposal-deposits are not considered in this ratio but are added to the absolute fields.

> This is a ratio of `(individualTotalPoolStake - proposal deposits)` to `(pdTotalActiveStake - proposal deposits)`

lehins replied:
No, please don't. We do not need to mention this, because this information about proposal deposits is only relevant for Conway and only for voting, so this will only be confusing for anyone who only cares about stake distribution with respect to leader election

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/PoolDistr.hs:
```diff
@@ -48,16 +54,21 @@ import NoThunks.Class (NoThunks (..))
+  encCBOR (IndividualPoolStake stake stakeCoin vrf) =
     mconcat
-      [ encodeListLen 2
+      [ encodeListLen 3 -- TODO: This needs versioning!
```
No versioning is needed, since it is not stored on chain. 
```suggestion
      [ encodeListLen 3
```
But it does need a corresponding change from `2` to `3` on the decoder: https://github.com/IntersectMBO/cardano-ledger/blob/c77828bcd0496faee4cb0b82a813f064f24c4465/libs/cardano-ledger-core/src/Cardano/Ledger/PoolDistr.hs#L77

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/PoolDistr.hs:
```diff
@@ -67,46 +78,70 @@ instance Crypto c => DecCBOR (IndividualPoolStake c) where
+
+instance Crypto c => EncCBOR (PoolDistr c) where
+  encCBOR (PoolDistr distr total) =
+    -- TODO: @aniketd: This needs versioning!
```
Ditto. No need. We only store in the ledger state
```suggestion
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/PoolDistr.hs:
```diff
@@ -67,46 +78,70 @@ instance Crypto c => DecCBOR (IndividualPoolStake c) where
+--     )
+--   where
+--   toBase (PoolDistr x) = x
+--   fromBase = PoolDistr
```
I think this is safe to remove.
```suggestion
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs:
```diff
@@ -291,6 +293,25 @@ umElemRDActive = \case
+data RewardDelegation c
+  = RewardDelegationSPO (KeyHash 'StakePool c) (CompactForm Coin)
+  | RewardDelegationDRep (DRep c) (CompactForm Coin)
+  | RewardDelegationBoth (KeyHash 'StakePool c) (DRep c) (CompactForm Coin)
```
Don't forget the strictness
```suggestion
  = RewardDelegationSPO !(KeyHash 'StakePool c) !(CompactForm Coin)
  | RewardDelegationDRep !(DRep c) !(CompactForm Coin)
  | RewardDelegationBoth !(KeyHash 'StakePool c) !(DRep c) !(CompactForm Coin)
```

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:
```diff
@@ -364,8 +364,8 @@ class (Show x, Adds x) => Sums t x | t -> x where
-  getSum (IndividualPoolStake r _) = r
-  genT _ r = IndividualPoolStake r <$> arbitrary
+  getSum (IndividualPoolStake r _ _) = r
+  genT _ r = IndividualPoolStake r mempty <$> arbitrary
```
Not quite sure if this is the correct approach, because there is a relation between `individualPoolStake`, `pdTotalActiveStake` and `individualTotalPoolStake`.

So it might be worth leaving a note about this in the comment.

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs:
```diff
@@ -291,6 +293,25 @@ umElemRDActive = \case
+-- We can tell that the pair is present and active when Txxxx has
+-- F's in the 1st, 3rd and 4th positions
+umElemDelegations :: UMElem c -> Maybe (RewardDelegation c)
+umElemDelegations = \case
```
This function needs a mention in the changelog

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs:
```diff
@@ -291,6 +293,25 @@ umElemRDActive = \case
+
+-- | Extract rewards that are either delegated to a DRep or an SPO (or both).
+-- We can tell that the pair is present and active when Txxxx has
+-- F's in the 1st, 3rd and 4th positions
```
```suggestion
-- F's in the 1st and either 3rd or 4th or both positions
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/EpochBoundary.hs:
```diff
@@ -102,8 +102,8 @@ instance Crypto c => DecShareCBOR (Stake c) where
-sumAllStake :: Stake c -> Coin
-sumAllStake = fromCompact . CompactCoin . VMap.foldl (\acc (CompactCoin c) -> acc + c) 0 . unStake
+sumAllStake :: Stake c -> CompactForm Coin
+sumAllStake = VMap.foldl (<>) mempty . unStake
```
I'd recommend adding `sumAllStakeCompact` that this function can use and `sumAllStake` can utilize as well.
This reduces boiler plate in functions that expect `Coin` and reduces breakage, while allowing you to use `sumAllStakeCompact` in the places where compact form is needed.

```suggestion
sumAllStake :: Stake c -> Coin
sumAllStake = fromCompact . CompactCoin . Compact

sumAllStakeCompact :: Stake c -> CompactForm Coin
sumAllStakeCompact = VMap.foldl (<>) mempty . unStake
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/EpochBoundary.hs:
```diff
@@ -297,17 +297,24 @@ calculatePoolDistr = calculatePoolDistr' (const True)
 calculatePoolDistr' :: forall c. (KeyHash 'StakePool c -> Bool) -> SnapShot c -> PoolDistr c
 calculatePoolDistr' includeHash (SnapShot stake delegs poolParams) =
-  let Coin total = sumAllStake stake
+  let total = sumAllStake stake
```
```suggestion
  let total = sumAllStakeCompact stake
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/EpochBoundary.hs:
```diff
@@ -297,17 +297,24 @@ calculatePoolDistr = calculatePoolDistr' (const True)
       -- total could be zero (in particular when shrinking)
       nonZeroTotal :: Integer
-      nonZeroTotal = if total == 0 then 1 else total
+      nonZeroTotal = if total == mempty then 1 else unCoin $ fromCompact total
```
```suggestion
      nonZeroTotalCompact = if total == mempty then 1 else total
      nonZeroTotalInteger = unCoin $ fromCompact nonZeroTotalCompact
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/EpochBoundary.hs:
```diff
@@ -297,17 +297,24 @@ calculatePoolDistr = calculatePoolDistr' (const True)
+            poolStakeMap
+            (VMap.toMap poolParams)
+        )
+        (if total == mempty then CompactCoin 1 else total)
```
Reduce duplication
```suggestion
        nonZeroTotalCompact
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/EpochBoundary.hs:
```diff
@@ -297,17 +297,24 @@ calculatePoolDistr = calculatePoolDistr' (const True)
+        ( Map.intersectionWith
+            ( \word64 poolparam ->
+                IndividualPoolStake
+                  (toInteger word64 % nonZeroTotal)
```
```suggestion
                  (toInteger word64 % nonZeroTotalInteger)
```

lehins commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/PoolLifetime.hs:
```diff
@@ -968,7 +980,7 @@ alicePerfEx11 = applyDecay decayFactor alicePerfEx8 <> epoch4Likelihood
     t = leaderProbability f relativeStake (unsafeBoundRational 0.5)
     -- everyone has delegated to Alice's Pool
-    Coin stake = EB.sumAllStake (EB.ssStake $ snapEx5 @c)
+    Coin stake = fromCompact $ EB.sumAllStake (EB.ssStake $ snapEx5 @c)
```
See my comment on `sumAllStake`
```suggestion
    Coin stake = EB.sumAllStake (EB.ssStake $ snapEx5 @c)
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState/PulsingReward.hs:
```diff
@@ -161,7 +161,7 @@ startStep slotsPerEpoch b@(BlocksMade b') es@(EpochState acnt ls ss nm) maxSuppl
           stakePerPool
           totalStake
-          activestake
+          $ UM.fromCompact activestake
```
See my comment on `sumAllStake`. If you can fix the camel case that'll be great too :slightly_smiling_face: 
```suggestion
          activeStake
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs:
```diff
@@ -160,43 +160,81 @@ instance EraPParams era => ToCBOR (PulsingSnapshot era) where
+  Map.foldlWithKey' go (dRepDistr, poolDistr) uMapChunk
   where
-    go accum stakeCred umElem =
+    go (drepAccum, poolAccum) stakeCred umElem =
```
Gotta avoid those space leaks. Tuples are lazy :wink: 
```suggestion
    go (!drepAccum, !poolAccum) stakeCred umElem =
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs:
```diff
@@ -160,43 +160,81 @@ instance EraPParams era => ToCBOR (PulsingSnapshot era) where
+        Just ips ->
+          distr
+            & poolDistrDistrL %~ Map.insert spo (ips & individualPoolStakeCoinL <>~ ccoin)
+            & poolDistrTotalL <>~ ccoin
```
We don't need to lookup or change the SPO distribution if the deposit is zero, which is going to be a 99% of the time.
```suggestion
    addToPoolDistr spo proposalDeposit distr = fromMaybe distr $ do
      guard (proposalDeposit /= mempty)
      ips <- Map.lookup spo $ distr ^. poolDistrDistrL
      pure (distr
            & poolDistrDistrL %~ Map.insert spo (ips & individualPoolStakeCoinL <>~ proposalDeposit)
            & poolDistrTotalL <>~ proposalDeposit
```

aniketd replied:
I thought that the comment would be helpful in making clear that even in the copy of `PoolDistr` that is part of the `DRepPulser` we do not adjust `individualPoolStake :: Rational` with proposal deposits. One can rely on that field being the same between both copies: the one in `NewEpochState` and the one in `DRepPulser`. What do you think? Is this wrong? Should that field be updated?

lehins replied:
This comment is misleading, because not only DRep pulser has nothing to do with rewards calculation, but also it is totally isolated and cannot interfere with anything except governance, i.e. RATIFY and ENACT rules.

> Should that field be updated?

Nope, leave it unchanged. There is no need to update something that is not used

lehins reviewed (commented):
The solution is solid. There are a few minor improvements that we can make, other than that this PR looks great.

aniketd committed:
Address review comments

aniketd commented on libs/cardano-ledger-core/src/Cardano/Ledger/EpochBoundary.hs:
```diff
@@ -22,6 +22,7 @@
 module Cardano.Ledger.EpochBoundary (
   Stake (..),
   sumAllStake,
+  sumAllStakeCompact,
```
TODO: Also add this to changelog. @aniketd

aniketd commented on libs/cardano-ledger-core/src/Cardano/Ledger/PoolDistr.hs:
```diff
@@ -48,65 +52,79 @@ import NoThunks.Class (NoThunks (..))
+-- necessary for the calculations in the `computeDRepDistr`.
+data PoolDistr c = PoolDistr
+  { unPoolDistr :: !(Map (KeyHash 'StakePool c) (IndividualPoolStake c))
+  , pdTotalActiveStake :: !(CompactForm Coin)
```
Add these changes to changelog @aniketd

aniketd commented on libs/cardano-ledger-core/src/Cardano/Ledger/PoolDistr.hs:
```diff
@@ -48,65 +52,79 @@ import NoThunks.Class (NoThunks (..))
   deriving anyclass (NFData, NoThunks)
 
+individualTotalPoolStakeL :: Lens' (IndividualPoolStake c) (CompactForm Coin)
+individualTotalPoolStakeL = lens individualTotalPoolStake $ \x y -> x {individualTotalPoolStake = y}
```
Add to changelog. @aniketd

aniketd commented on libs/cardano-ledger-core/src/Cardano/Ledger/PoolDistr.hs:
```diff
@@ -48,65 +52,79 @@ import NoThunks.Class (NoThunks (..))
-  where
-  toExp (PoolDistr x) = Base MapR x
+poolDistrTotalL :: Lens' (PoolDistr c) (CompactForm Coin)
+poolDistrTotalL = lens pdTotalActiveStake $ \x y -> x {pdTotalActiveStake = y}
```
To changelog @aniketd

aniketd committed:
Add more details to changelog

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs:
```diff
@@ -160,43 +161,81 @@ instance EraPParams era => ToCBOR (PulsingSnapshot era) where
+      -- and denominator in order not to interfere with rewards calculation,
+      -- although this is an isolated copy inside the DRepPulser, we want to
+      -- just be sure we do not confuse ourselves over which representations are
+      -- used where.
```
This comment is totally unnecessary
```suggestion
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs:
```diff
@@ -320,11 +365,12 @@ finishDRepPulser (DRPulsing (DRepPulser {..})) =
   where
     !leftOver = Map.drop dpIndex $ umElems dpUMap
-    !finalDRepDistr = computeDRepDistr dpStakeDistr dpDRepState dpProposalDeposits dpDRepDistr leftOver
+    (!finalDRepDistr, !finalStakePoolDistr) =
```
Those bangs aren't really necessary, since `ratifyEnv` will force them, but if you do want to add them then you also need to add a bang on the let binding itself, since those are lazy.
```suggestion
    !(!finalDRepDistr, !finalStakePoolDistr) =
```

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Vars.hs:
```diff
@@ -680,7 +680,7 @@ markPoolDistrL ::
 
 pooldistrHelpL :: Lens' (PoolDistr c) (Map (KeyHash 'StakePool c) (IndividualPoolStake c))
-pooldistrHelpL = lens unPoolDistr (\_ u -> PoolDistr u)
+pooldistrHelpL = lens unPoolDistr (\x y -> x {unPoolDistr = y})
```
Remove this lens in favor of the newly added `poolDistrDistrL`. No need to keep duplicate code around
```suggestion
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs:
```diff
@@ -291,6 +293,32 @@ umElemRDActive = \case
+
+-- | Extract rewards that are either delegated to a DRep or an SPO (or both).
+-- We can tell that the pair is present and active when Txxxx has F's in the 1st
+-- and either 3rd or 4th or both positions. If there are no rewards or deposits
+-- but the delegations still exist, then we return zero coin as reward.
```
Very nice! I didn't even think about this case!

lehins commented on libs/cardano-ledger-core/CHANGELOG.md:
```diff
@@ -2,6 +2,14 @@
 ## 1.12.0.0
 
+* Add `UMap.umElemDelegations` to extract SPO, DRep and rewards for a given stake credential. #4324
+  * Add `sumAllStakeCompact` and redefine `sumAllStake` as a composition of it.
```
That is an irrelevant information in the changelog. Commit message maybe, but not for the changelog, since it is not a visible change
```suggestion
* Add `sumAllStakeCompact`
```

aniketd committed:
Address another set of review comments

aniketd committed:
Fix tests in cardano-ledger-shelley-test.

We reset the newly added fields: individualTotalPoolStake and
pdTotalActiveStake, in the ledger state that testCHAINEXample compares
with checkTrace, so as to not interfere with the chained ledger states
in these tests.

aniketd committed:
Fix the zero denominator issue with PoolDistr.

The prop_RATIFY test revealed a bug in the current implementation of
spoAcceptedRatio that constructs a `Ratio`  out of the numerator and
denominator of the PoolDistr. We use the same method of defaulting it to
unit if it is zero.

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs:
```diff
@@ -199,28 +198,29 @@ spoAccepted re rs gas =
 -- vote are considered as `Abstain` votes.
 spoAcceptedRatio :: forall era. RatifyEnv era -> GovActionState era -> Rational
 spoAcceptedRatio
```
As we've discussed on the call, if you can make this function look a lot more like `drepAcceptedRatio` this PR will be good to go

lehins reviewed (commented):
PR looks great. Just one comment that we've discussed on a call

aniketd committed:
Make spoAcceptedRatio read like drepAcceptedRatio

Soupstraw commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
```diff
@@ -796,14 +796,125 @@ votingSpec =
           -- for DRep votes to ratify a proposal.
           passNEpochs 2
           getLastEnactedCommittee `shouldReturn` SJust (GovPurposeId addCCGaid)
+        describe "Proposal deposits contribute to active voting stake" $ do
```
I think there should also be tests for delegation of deposit to `DRep`s

aniketd replied:
Yes that is covered in a previous PR. The tests are present in the same module under the DReps hierarchy.

Soupstraw commented on libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs:
```diff
@@ -291,6 +293,32 @@ umElemRDActive = \case
+  TEFFE _ spo -> Just $ RewardDelegationSPO spo mempty
+  TEFFF _ spo drep -> Just $ RewardDelegationBoth spo drep mempty
+  _ -> Nothing
+{-# INLINE umElemDelegations #-}
```
I think it'd look nicer if you used functions to access each field instead of doing a direct pattern match on the ugly `UMElem` type.

Soupstraw replied:
so something like
```
case (umElemRDPair x, umElemSPool x, umElemDRep x) of
  ...
```

aniketd replied:
Sorry, I could not find an `ElemP` anywhere, if you mean `umElemAsTuple`, would that not be a bit finicky to deal with all the `StrictMaybe`s inside of the tuple?

Soupstraw replied:
It seems less finicky than dealing with all of the combinations that you're pattern matching on right now. For example 
```
  TFEFE RDPair {rdReward} spo -> Just $ RewardDelegationSPO spo rdReward
  TFFFE RDPair {rdReward} _ spo -> Just $ RewardDelegationSPO spo rdReward
```
  would become just a single branch
```
  (SJust (RDPair {rdReward}), SJust spo, SNothing) -> Just $ RewardDelegationSPO spo rdReward
```
which reduces code repetition.

Soupstraw replied:
I think instead of the tuple you can just use the `UMElem` pattern

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs:
```diff
@@ -183,44 +181,34 @@ spoAccepted re rs gas =
     }
-    | abstainVotesRatio == 1 = 0 -- guard against the degenerate case when all abstain.
-    | otherwise = yesVotesRatio / (1 - abstainVotesRatio)
+    | totalActiveStake == 0 = 0 -- guard against the degenerate case when active stake is zero.
```
I'm not sure this is necessary, because the next case pretty much implies this anyways, but it doesn't hurt either to have it here.

aniketd replied:
Yes, I have just added both :+1:

Soupstraw commented on libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs:
```diff
@@ -291,6 +293,32 @@ umElemRDActive = \case
+  TEFFF _ spo drep -> Just $ RewardDelegationBoth spo drep mempty
+  _ -> Nothing
+{-# INLINE umElemDelegations #-}
+
```
Unrelated to this PR, but this file seems to have a couple of comments referring to an `ElemP` pattern which doesn't seem to exist anymore. We should probably update these comments.

I think it's been renamed to `UMElem`

aniketd replied:
Oh got it! I will remove those comments then :+1:  Thanks! :raised_hands: @Soupstraw I need an approval on this PR. Alexey has already reviewed it :+1:

aniketd committed:
Fix mention of pattern ElemP to UMElem for UMap

Soupstraw reviewed (approved):
LGTM!

aniketd committed:
Make umElemDelegations better
